Issue #1981: WIP MultiProvider RC extension

abdulazizali77 opened the pull request:
@gvagenas @deruelle  I thought about possible ways to make account specific mappings to RC configurations with the currect RC Extensions API. 

Abstractly i believe any plugin/extension API should allow its consumers to modify aspects of the system at predefined levels/modules. In reference to SMSC MProc API, it allows Messages and anything related to the Message, network id etc to be customized. 
I believe with RC we can abstract that out even more, from Message to Session to Transaction to SystemNode to System, which is why i decided to subclass the `ExtensionsResponse` as i did. 

Implementation wise, we couldve chosen between two options
1. Pass in an/several Interface/s to the Extension handler interface and expose certain parts of our System/Node/Tx/Session/Message to be customized by the extension provider/implementor. 
This is sort of how SMSC does it, it passes specific factories in specific handlers which expose specific configurations to be modified by the `MProcRule` implementations.
2. Return specific class/Interface/es to the calling method and handle appropriately. The class/interface would be subclasses/subinterfaces of the `ExtensionResponse` and RC would handle them all differently.

In the PR changes ive opted to go with 2) so we can leverage the existing `ExtensionResponse`. 
Additionally we also need to pass in more information to the extension than just a message object, because in our example of the MultiProvider specification it needs to map Account to custom database configurations. So abstractly we also need to create a new class/interface to pass in to the handlers, maybe something like `ExtensionRequest` ?
What do you think? If you think this is in sort of the correct direction, i will proceed with defining an `ExtensionRequest` and writing the actual MultiProvider extension itself.

deruelle commented:
@abdulazizali77 that seems to make sense to me for defining multiple layers of Response but I think we should really describe what each level means in the context of Restcomm so it's clear and documentation be updated on that as well. 

Seems to Makes sense to provide an ExtensionRequest as well to me. 

@gvagenas will take the  final call on this as he has more insights.

scottbarstow commented on restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/smpp/SmppMessageHandler.java:
```diff
@@ -33,18 +36,22 @@
 import java.util.List;
 import java.util.concurrent.TimeUnit;
 
+import javolution.util.FastList;
```
@abdulazizali77 this looks like a new dependency. Not sure what policy is, just want to raise the issue here.

abdulazizali77 replied:
@scottbarstow unsure myself, should find out. Smsc common-library is dependent on this library so license *MIGHT* be permissible. 
However we can cast this to `Collection<Tlv>` .

scottbarstow commented on restcomm/restcomm.dao/src/main/java/org/restcomm/connect/dao/entities/SmsMessage.java:
```diff
@@ -47,13 +46,20 @@
     private final Direction direction;
     private final BigDecimal price;
     private final Currency priceUnit;
+    private final String tlvTag;
```
Should we support array here, like TlvSet, rather than just a single key/value pair? @abdulazizali77 what do you think?

abdulazizali77 replied:
@scottbarstow I do agree, currently as per your comment on Slack the SmsService only fetches tlv 1550 which is why its only adding one tlv. However the code youre pointing at is the for-test-only Sms REST API tlv modification which im going to clean out.

scottbarstow commented on restcomm/restcomm.application/src/main/webapp/WEB-INF/data/hsql/restcomm.script:
```diff
@@ -21,12 +21,15 @@ CREATE MEMORY TABLE "restcomm_gateways"("sid" VARCHAR(34) NOT NULL PRIMARY KEY,"
 SET WRITE_DELAY 10
 SET SCHEMA PUBLIC
 INSERT INTO "restcomm_accounts" VALUES('ACae6e420f425248d6a26948c17a9e2acf','2012-04-24 00:00:00.000000000','2012-04-24 00:00:00.000000000','administrator@company.com','Default Administrator Account',NULL,'Full','uninitialized','77f8c12cc7b8f8423e5c38b035249166','Administrator','/2012-04-24/Accounts/ACae6e420f425248d6a26948c17a9e2acf')
+INSERT INTO "restcomm_accounts" VALUES('ACae6e420f425248d6a26948c17a9e2acg','2012-04-24 00:00:00.000000000','2012-04-24 00:00:00.000000000','second@company.com','Second Administrator Account',NULL,'Full','uninitialized','77f8c12cc7b8f8423e5c38b035249166','Administrator','/2012-04-24/Accounts/ACae6e420f425248d6a26948c17a9e2acg')
```
Doubt we want to add new default admin accounts just to demonstrate the capabilities of this particular feature. Should have George weigh in on this for sure.

abdulazizali77 replied:
@scottbarstow agreed, this was only for test purposes

scottbarstow commented:
@abdulazizali77 are we ready for George to review this PR yet?

abdulazizali77 commented:
@gvagenas This PR is ready. Please have a look,

maria-farooq commented on restcomm/restcomm.application/src/main/webapp/WEB-INF/scripts/mariadb/init.sql:
```diff
@@ -381,6 +381,13 @@ date_updated DATETIME,
+CREATE TABLE restcomm_accounts_extensions (
+account_sid VARCHAR(34) NOT NULL,
+extension_sid VARCHAR(34) NOT NULL,
+extension_blob CLOB,
```
@abdulazizali77 is it a typing mistake? `CLOB`

maria-farooq commented:
@abdulazizali77 we need to add upgrade scripts in bitbucket RC to accomodate db schema changes. please check `https://bitbucket.org/telestax/telscale-restcomm/src/master/restcomm/configuration/config-scripts/as7-config-scripts/restcomm/upgrade_scripts/?at=master`

deruelle commented:
@gvagenas @abdulazizali77 can this be merged ?

gvagenas commented on restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/SmsService.java:
```diff
@@ -208,8 +218,8 @@ private void message(final Object message) throws IOException {
                 trying.send();
-
-                ActorRef session = session();
+                //TODO:do extensions check here too?
```
@abdulazizali77 yes we need extension check here also. Those requests are MESSAGE requests from registered clients that Restcomm will proxy out to.

abdulazizali77 replied:
@gvagenas I want to address this in a new Issue/PR because this Use Case was not actually covered in our testing.

gvagenas commented on restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/SmsSession.java:
```diff
@@ -113,6 +120,18 @@ public SmsSession(final Configuration configuration, final SipFactory factory, f
+                //FIXME:fetch the tag value from the arbitrary key
+                this.tlvSet.addOptionalParameter(new Tlv(SmppConstants.TAG_DEST_NETWORK_ID,ByteArrayUtil.toByteArray(Integer.parseInt(valStr))));
+            } catch (Exception e) {
+                // TODO: handle exception
```
@abdulazizali77 please add a logger.error(...) here

abdulazizali77 replied:
@gvagenas added.

gvagenas commented on restcomm/restcomm.http/src/main/java/org/restcomm/connect/http/ExtensionsConfigurationEndpoint.java:
```diff
@@ -87,23 +88,44 @@ void init() {
+            }
+        }
+
+        if (accountSid!=null) {
```
@abdulazizali77 I don't think this is proper here. As far as I understand, an extension can have a global configuration but also can have a configuration specific to an account. Thus we need to different methods, one to get the global configuration and one to get the configuration for a specific account

abdulazizali77 replied:
@gvagenas I want to address the REST design + implementation issues in #2127

gvagenas commented on restcomm/restcomm.dao/src/main/java/org/restcomm/connect/dao/ExtensionsConfigurationDao.java:
```diff
@@ -97,4 +97,16 @@
     boolean validate(ExtensionConfiguration extensionConfiguration);
+
+    /**/
+    ExtensionConfiguration getAccountExtensionConfiguration(String accountSid, String extensionSid);
```
@abdulazizali77 please add short description as javadoc comment to each one of the new methods here.

abdulazizali77 replied:
@gvagenas added

gvagenas commented on restcomm/restcomm.dao/src/main/java/org/restcomm/connect/dao/mybatis/MybatisExtensionsConfigurationDao.java:
```diff
@@ -276,4 +283,86 @@ private ExtensionConfiguration toExtensionConfiguration(final Map<String, Object
+                    map.put("account_sid", DaoUtils.writeSid(accountSid));
+                    map.put("extension_sid", DaoUtils.writeSid(extensionConfiguration.getSid()));
+
+                    if (extensionConfiguration.getConfigurationData() != null)
```
@abdulazizali77 I believe that first thing we should do is to check that configuration data is not null and then proceed, if data is null we shouldn't attempt to do an insert in the DB. 
Also, the check should be done at ExtensionsConfigurationEndpoint for add and update operations

abdulazizali77 replied:
@gvagenas I want to address the REST design + implementation issues in #2127

abdulazizali77 replied:
@gvagenas Please have another look.
-I have addressed most of the comments
-I have added the common Extension handler functions to ExtensionController but have not done this for CallManager.
-I have moved also `DefaultExtensionConfiguration` from telscale to community
-I have deferred the REST design + implementation issues to be addressed in #2127. I think we should discuss about the design at first.
-I have deferred from adding the extension hook for "MESSAGE" use case as this seems to be a new scenario which was not even encountered during testing. Will open an Issue for this

abdulazizali77 commented on restcomm/restcomm.application/src/main/webapp/WEB-INF/data/hsql/restcomm.script:
```diff
@@ -42,3 +45,7 @@ INSERT INTO "restcomm_incoming_phone_numbers" VALUES('PN46678e5b01d44973bf184f65
 INSERT INTO "restcomm_incoming_phone_numbers" VALUES('PNb43ed9e641364277b6432547ff1109e9','2014-02-17 22:37:19.392000000','2014-02-17 22:37:19.392000000','RVD external services app, customer ID 1 or 2 ','ACae6e420f425248d6a26948c17a9e2acf','+1241','2012-04-24',FALSE,NULL,'POST',NULL,'POST',NULL,'POST','AP81cf45088cba4abcac1261385916d582',NULL,'POST',NULL,'POST',NULL,'/2012-04-24/Accounts/ACae6e42 ...
 INSERT INTO "restcomm_clients" VALUES('CLa2b99142e111427fbb489c3de357f60a','2013-11-04 12:52:44.144000000','2013-11-04 12:52:44.144000000','ACae6e420f425248d6a26948c17a9e2acf','2012-04-24','alice','alice','1234',1,NULL,'POST',NULL,'POST',NULL,'/2012-04-24/Accounts/ACae6e420f425248d6a26948c17a9e2acf/Clients/CLa2b99142e111427fbb489c3de357f60a')
 INSERT INTO "restcomm_clients" VALUES('CL3003328d0de04ba68f38de85b732ed56','2013-11-04 16:33:39.248000000','2013-11-04 16:33:39.248000000','ACae6e420f425248d6a26948c17a9e2acf','2012-04-24','bob','bob','1234',1,NULL,'POST',NULL,'POST',NULL,'/2012-04-24/Accounts/ACae6e420f425248d6a26948c17a9e2acf/Clients/CL3003328d0de04ba68f38de85b732ed56')
+INSERT INTO "restcomm_accounts_extensions" VALUES('ACae6e420f425248d6a26948c17a9e2acf','EX00000000000000000000000000000005','{"outbound-sms": [{"destination_network_id": "1000"}]}')
```
@gvagenas removed

abdulazizali77 commented on restcomm/restcomm.application/pom.xml:
```diff
@@ -272,7 +272,11 @@
-
+        <dependency>
+            <groupId>org.apache.maven</groupId>
+            <artifactId>maven-artifact</artifactId>
```
@gvagenas removed, it was initially because the ExtensionConfiguration requires it

abdulazizali77 commented on restcomm/restcomm.dao/src/main/java/org/restcomm/connect/dao/ExtensionsConfigurationDao.java:
```diff
@@ -97,4 +97,16 @@
      */
     boolean validate(ExtensionConfiguration extensionConfiguration);
+
+    /**/
```
@gvagenas added

abdulazizali77 commented on restcomm/restcomm.extension.api/src/main/java/org/restcomm/connect/extension/api/MessageExtensionResponse.java:
```diff
@@ -0,0 +1,3 @@
+package org.restcomm.connect.extension.api;
+public class MessageExtensionResponse extends ExtensionResponse {
```
@gvagenas added

abdulazizali77 commented on restcomm/restcomm.extension.api/src/main/java/org/restcomm/connect/extension/api/NodeExtensionResponse.java:
```diff
@@ -0,0 +1,3 @@
+package org.restcomm.connect.extension.api;
+public class NodeExtensionResponse extends ExtensionResponse {
```
@gvagenas added

abdulazizali77 commented on restcomm/restcomm.extension.api/src/main/java/org/restcomm/connect/extension/api/SessionExtensionResponse.java:
```diff
@@ -0,0 +1,11 @@
+package org.restcomm.connect.extension.api;
+import org.apache.commons.configuration.Configuration;
+public class SessionExtensionResponse extends ExtensionResponse {
```
@gvagenas added
I think the design of Extensions, the exposed API, ExtensionsResponse is something we still need to discuss.
Initially i think of `SessionResponse` as Sms agnostic, for me i guess a running Call might be defined a 'Session'

abdulazizali77 commented on restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/SmsService.java:
```diff
@@ -325,32 +337,79 @@ private boolean redirectToHostedSmsApp(final ActorRef self, final SipServletRequ
         if (extensions != null && extensions.size() > 0) {
+
             for (RestcommExtensionGeneric extension : extensions) {
+                logger.info( "isEnabled="+extension.isEnabled());
```
@gvagenas added.

abdulazizali77 commented on restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/SmsSession.java:
```diff
@@ -135,7 +154,8 @@ private void inbound(final Object message) throws IOException {
             // Store the last sms event.
-            last = new SmsSessionRequest(from, to, body, customRequestHeaderMap);
+
+            last = new SmsSessionRequest(from, to, body, Encoding.GSM, customRequestHeaderMap, this.tlvSet);
```
@gvagenas removed.
I have added extra constructors to `SmsSessionRequest`. Additionally i believe it needs further cleanup

abdulazizali77 commented on restcomm/restcomm.application/src/main/webapp/WEB-INF/conf/extensions.xml:
```diff
@@ -11,11 +11,9 @@
-    <extension>
-        <name>ExampleExtension</name>
-        <class>org.restcomm.connect.ExampleExtension</class>
+    <!-- extension name="MultiProvider">
```
@gvagenas removed. I have also added the correct structure.

abdulazizali77 commented:
@gvagenas Please have another look.
-I have addressed most of the comments
-I have added the common Extension handler functions to ExtensionController but have not done this for CallManager.
-I have moved also DefaultExtensionConfiguration from telscale to community
-I have deferred the REST design + implementation issues to be addressed in #2127. I think we should discuss about the design at first.
-I have deferred from adding the extension hook for "MESSAGE" use case as this seems to be a new scenario which was not even encountered during testing. Will open an Issue for this

abdulazizali77 commented:
thanks @gvagenas
